orchestrator/src/worker.rs

```diff
@@ -0,0 +1,656 @@
+                                .await
+                                .context(UnableToSendWriteFileResponseSnafu)
+                        });
+                    }
+
+                    CoordinatorMessage::ReadFile(req) => {
+                        let project_dir = project_dir.clone();
+                        let worker_msg_tx = worker_msg_tx();
+
+                        tasks.spawn(async move {
+                            worker_msg_tx
+                                .send(handle_read_file(req, project_dir).await)
+                                .await
+                                .context(UnableToSendReadFileResponseSnafu)
+                        });
+                    }
+
+                    CoordinatorMessage::ExecuteCommand(req) => {
+                        cmd_tx
+                            .send((Multiplexed(job_id, req), worker_msg_tx()))
+                            .await
+                            .drop_error_details()
+                            .context(UnableToSendCommandExecutionRequestSnafu)?;
+                    }
+
+                    CoordinatorMessage::StdinPacket(data) => {
+                        stdin_tx
+                            .send(Multiplexed(job_id, data))
+                            .await
+                            .drop_error_details()
+                            .context(UnableToSendStdinPacketSnafu)?;
+                    }
+                }
+            }
+
+            Some(task) = tasks.join_next() => {
+                task.context(TaskExitedSnafu)??;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub enum HandleCoordinatorMessageError {
+    UnableToSendWriteFileResponse {
+        source: MultiplexingSenderError,
+    },
+
+    UnableToSendReadFileResponse {
+        source: MultiplexingSenderError,
+    },
+
+    #[snafu(display("Failed to send command execution request"))]
+    UnableToSendCommandExecutionRequest {
+        source: mpsc::error::SendError<()>,
+    },
+
+    #[snafu(display("Failed to send stdin packet"))]
+    UnableToSendStdinPacket {
+        source: mpsc::error::SendError<()>,
+    },
+
+    TaskExited {
+        source: tokio::task::JoinError,
+    },
+}
+
+#[derive(Debug, Clone)]
+struct MultiplexingSender {
+    job_id: JobId,
+    tx: mpsc::Sender<Multiplexed<WorkerMessage>>,
+}
+
+impl MultiplexingSender {
+    async fn send(
+        &self,
+        message: Result<impl Into<WorkerMessage>, impl std::error::Error>,
+    ) -> Result<(), MultiplexingSenderError> {
+        match message {
+            Ok(v) => self.send_ok(v).await,
+            Err(e) => self.send_err(e).await,
+        }
+    }
+
+    async fn send_ok(
+        &self,
+        message: impl Into<WorkerMessage>,
+    ) -> Result<(), MultiplexingSenderError> {
+        self.send_raw(message.into()).await
+    }
+
+    async fn send_err(&self, e: impl std::error::Error) -> Result<(), MultiplexingSenderError> {
```

Comment:
```suggestion
    async fn send_err(&self, e: impl snafu::Error) -> Result<(), MultiplexingSenderError> {
```

This is basically the same, but a *tiny* bit better because when `core::Error` is stabilized this will be able to change transparently. (and I used `snafu::Error` elsewhere, so this is for consistency)